src/webgpu/api/operation/render_pipeline/primitive_topology.spec.ts

```diff
@@ -0,0 +1,376 @@
+        color: expectedColor,
+      },
+      {
+        // Triangle {v3, v4, v5}
+        location: Point2D.getCentroid(VertexLocations[2], VertexLocations[3], VertexLocations[4]),
+        color: expectedColor,
+      },
+    ];
+  }
+
+  run(): void {
+    const colorAttachment = this.makeAttachmentTexture();
+
+    // Color load operator will clear color attachment to zero.
+    const encoder = this.device.createCommandEncoder();
+    const renderPass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: colorAttachment.createView(),
+          loadValue: { r: 0.0, g: 0.0, b: 0.0, a: 0.0 },
+        },
+      ],
+    });
+
+    // Draw a primitive using 6 vertices based on the type.
+    // Pixels are generated based on vertex position.
+    // If point, 1 pixel is generated at each vertex location.
+    // Otherwise, >1 pixels could be generated.
+    const vertexModule = this.makeShaderModule('vertex', {
+      glsl: `#version 450
+            layout(location = 0) in vec4 pos;
+            void main() {
+                gl_Position = pos;
+                gl_PointSize = 1.0;
+            }`,
+    });
+
+    // Output color is solid green.
+    const fragmentModule = this.makeShaderModule('fragment', {
+      glsl: `#version 450
+            layout(location = 0) out vec4 fragColor;
+            void main() {
+              fragColor = vec4(0.0, 1.0, 0.0, 1.0);
+            }`,
+    });
+
+    renderPass.setPipeline(
+      this.device.createRenderPipeline({
+        vertexStage: { module: vertexModule, entryPoint: 'main' },
+        fragmentStage: { module: fragmentModule, entryPoint: 'main' },
+        primitiveTopology: this.params.primitiveTopology,
+        rasterizationState: {
+          frontFace: 'ccw',
+          cullMode: 'none',
+        },
+        colorStates: [{ format: kColorFormat }],
+        vertexState: {
+          vertexBuffers: [
+            {
+              arrayStride: 4 * Float32Array.BYTES_PER_ELEMENT,
+              attributes: [
+                {
+                  format: 'float4',
+                  offset: 0,
+                  shaderLocation: 0,
+                },
+              ],
+            },
+          ],
+        },
+      })
+    );
+
+    // Create vertices for the primitive.
+    const vertexCoords = PrimitiveTopologyTest.generateVertexBuffer(VertexLocations);
+    const vertexBuffer = this.makeVertexBuffer(vertexCoords);
+
+    renderPass.setVertexBuffer(0, vertexBuffer);
+    renderPass.draw(6, 1, 0, 0);
+    renderPass.endPass();
+
+    this.device.defaultQueue.submit([encoder.finish()]);
+
+    // Compute test locations for valid and invalid pixels.
+    // If the primitive covers the pixel, the color value will be |kValidPixelColor|.
+    // Otherwise, a non-covered pixel will be |kInvalidPixelColor|.
+    let testLocations: TestLocation[] = [];
+    switch (this.params.primitiveTopology) {
```

Comment:
nit: It's unfortunate to have so much of the test embedded into this switch statement. Perhaps instead of this with the parameterized single test, we could have one test for each primitive topology, and each test could just embed the contents of these switch cases, and pass it into `t.run(primitiveTopology, testLocations)`. WDYT?